### PR TITLE
[WIP] Some minor changes to HDFS and to docs about globs

### PR DESCRIPTION
**Pull request by stavrolia**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Minor improvement

### CODE REVIEW

**Comment by stavrolia** [comment]
For #5371 issue.

File: dbms/src/IO/WriteBufferFromHDFS.cpp [hunk 1]
```diff
@@ -32,9 +33,12 @@ struct WriteBufferFromHDFS::WriteBufferFromHDFSImpl
     {
         const size_t begin_of_path = hdfs_uri.find('/', hdfs_uri.find("//") + 2);
         const std::string path = hdfs_uri.substr(begin_of_path);
-        if (path.find("*?{") != std::string::npos)
+        if (path.find("^{*") != std::string::npos)
```
Comment (alexey-milovidov) [inline dbms/src/IO/WriteBufferFromHDFS.cpp]: What is the reason for this change?

Comment (alexey-milovidov) [reply dbms/src/IO/WriteBufferFromHDFS.cpp]: And it looks totally incorrect because it will search for substring.
Do we have a test for this exception?

File: docs/en/operations/table_engines/hdfs.md [hunk 2]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
     - Indexes.
     - Replication.
 
+**Globs in path**
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
## Globs in Path
```

File: docs/en/operations/table_engines/hdfs.md [hunk 3]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+
+Multiple path components can have globs. For being processed file should exists and matches to the whole path pattern. Listing of files determines during `SELECT` (not at `CREATE` moment).
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
Multiple path components can have globs. File should exist and match the whole path pattern to be processed. List of available files is determined during each `SELECT` (not at the moment of `CREATE TABLE`).
```

File: docs/en/operations/table_engines/hdfs.md [hunk 4]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+
+- `*` — Substitutes any number of any characters including none.
+- `?` — Substitutes any single character.
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]: would be useful to mention if it supports unicode or it's only ascii characters

File: docs/en/operations/table_engines/hdfs.md [hunk 5]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+- `{some_string,another_string,yet_another_one}` — Substitutes any of strings `'some_string', 'another_string', 'yet_another_one'` (similar to the remote table function).
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]: remote table function -> link to it's docs

File: docs/en/operations/table_engines/hdfs.md [hunk 6]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+- `{N..M}` — Substitutes any number in range from N to M including both borders (similar to the [remote table function]()).
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]: missing link

File: docs/en/operations/table_engines/hdfs.md [hunk 7]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+
+**Example**
+
+1. Suppose that we have several files in TSV format with following URIs on HDFS:
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
1. Suppose we have several files in TSV format with the following URIs on HDFS:
```

File: docs/en/operations/table_engines/hdfs.md [hunk 8]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+
+- 'hdfs://hdfs1:9000/some_dir/some_file_1'
+- 'hdfs://hdfs1:9000/some_dir/some_file_2'
+- 'hdfs://hdfs1:9000/some_dir/some_file_3'
+- 'hdfs://hdfs1:9000/another_dir/some_file_1'
+- 'hdfs://hdfs1:9000/another_dir/some_file_2'
+- 'hdfs://hdfs1:9000/another_dir/some_file_3'
+
+2. There are several ways to make table consists of all six files:
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
2. There are several ways to make a table consisting of all six files:
```

File: docs/en/operations/table_engines/hdfs.md [hunk 9]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+CREATE TABLE table_with_range (name String, value UInt32) ENGINE = HDFS('hdfs://hdfs1:9000/{some,another}_dir/some_file_{1..3}', 'TSV')
+```
+
+Another way:
+
+```sql
+CREATE TABLE table_with_question_mark (name String, value UInt32) ENGINE = HDFS('hdfs://hdfs1:9000/{some,another}_dir/some_file_?', 'TSV')
+```
+
+Table consists of all the files in both directories (all files should satisfy format and schema described in query):
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
Table consisting of all the files in both directories (all files should satisfy format and schema described in query):
```

File: docs/en/operations/table_engines/hdfs.md [hunk 10]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+
+```sql
+CREATE TABLE table_with_asterisk (name String, value UInt32) ENGINE = HDFS('hdfs://hdfs1:9000/{some,another}_dir/*', 'TSV')
+```
+
+!!! warning
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
!!! warning "Warning"
```

File: docs/en/operations/table_engines/hdfs.md [hunk 11]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+    If the listing of files contains number ranges with leading zeros, use the construction with braces for each digit separately or use `?`.
+
+**Example**
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]: we should probably ident this, so example will be inside the "warning" block, not after it (not sure though)

File: docs/en/operations/table_engines/hdfs.md [hunk 12]
```diff
@@ -48,4 +49,53 @@ SELECT * FROM hdfs_engine_table LIMIT 2
...
+
+Create table with files named `file000`, `file001`, ... , `file999`:
+
+```sql
+CREARE TABLE big_table (name String, value UInt32) ENGINE = HDFS('hdfs://hdfs1:9000/big_dir/file{0..9}{0..9}{0..9}', 'CSV')
```
Comment (blinkov) [inline docs/en/operations/table_engines/hdfs.md]:
```suggestion
CREARE TABLE big_table (name String, value UInt32)
ENGINE = HDFS('hdfs://hdfs1:9000/big_dir/file{0..9}{0..9}{0..9}', 'CSV')
```

I'd add line break before `ENGINE =` instead of after in previous examples as well.

File: docs/en/query_language/table_functions/file.md [hunk 13]
```diff
@@ -1,7 +1,7 @@
 
 # file
 
-Creates a table from a file.
+Creates a table from a file. This table function is similar to [hdfs](hdfs.md) one.
```
Comment (blinkov) [inline docs/en/query_language/table_functions/file.md]:
```suggestion
Creates a table from a file. This table function is similar to [HDFS](hdfs.md) one.
```

what about url?

**Review by blinkov — CHANGES_REQUESTED** [review]

File: docs/en/query_language/table_functions/file.md (hunk 13 above)
Comment (stavrolia) [reply docs/en/query_language/table_functions/file.md]: Not so similar because of globs, but I would add it.

Table function hdfs named hdfs, not HDFS, do we need capital letters here?

Comment (blinkov) [reply docs/en/query_language/table_functions/file.md]: > Table function hdfs named hdfs, not HDFS, do we need capital letters here?

@stavrolia probably not, but that looks weird and might be confusing, probably worth to make it case-insensitive.

File: docs/en/operations/table_engines/hdfs.md (hunk 11 above)
Comment (stavrolia) [reply docs/en/operations/table_engines/hdfs.md]: I think there is no matter to be inside or after "warning" block

**Comment by stavrolia** [comment]
This PR moved to  #7016
